relay/polkadot/src/coretime_migration.rs

```diff
@@ -0,0 +1,319 @@
+				err
+			);
+		}
+	}
+
+	let config = configuration::ActiveConfig::<T>::get();
+	// num_cores was on_demand_cores until now:
+	for on_demand in 0..config.scheduler_params.num_cores {
+		let core = CoreIndex(legacy_count.saturating_add(on_demand as _));
+		let r = parachains_assigner_coretime::Pallet::<T>::assign_core(
+			core,
+			now,
+			vec![(CoreAssignment::Pool, PartsOf57600::FULL)],
+			None,
+		);
+		if let Err(err) = r {
+			log::error!("Creating assignment for existing on-demand core, failed: {:?}", err);
+		}
+	}
+	let total_cores = config.scheduler_params.num_cores + legacy_count;
+	configuration::ActiveConfig::<T>::mutate(|c| {
+		c.scheduler_params.num_cores = total_cores;
+	});
+
+	if let Err(err) = migrate_send_assignments_to_coretime_chain::<T, SendXcm, LegacyLease>() {
+		log::error!("Sending legacy chain data to coretime chain failed: {:?}", err);
+	}
+
+	let single_weight = <T as Config>::WeightInfo::assign_core(1);
+	single_weight
+		.saturating_mul(u64::from(legacy_count.saturating_add(config.scheduler_params.num_cores)))
+		// Second read from sending assignments to the coretime chain.
+		.saturating_add(T::DbWeight::get().reads_writes(2, 1))
+}
+
+fn migrate_send_assignments_to_coretime_chain<
+	T: Config,
+	SendXcm: xcm::v4::SendXcm,
+	LegacyLease: GetLegacyLease<BlockNumberFor<T>>,
+>() -> result::Result<(), SendError> {
+	let legacy_paras = LegacyLease::get_all_parachains_with_leases();
+	let legacy_paras_count = legacy_paras.len();
+	let (system_chains, lease_holding): (Vec<_>, Vec<_>) =
+		legacy_paras.into_iter().partition(IsSystem::is_system);
+
+	let reservations = system_chains.into_iter().map(|p| {
+		let schedule = BoundedVec::truncate_from(vec![ScheduleItem {
+			mask: CoreMask::complete(),
+			assignment: CoreAssignment::Task(p.into()),
+		}]);
+		mk_coretime_call::<T>(CoretimeCalls::Reserve(schedule))
+	});
+
+	let mut leases = lease_holding.into_iter().filter_map(|p| {
+			log::trace!(target: "coretime-migration", "Preparing sending of lease holding para {:?}", p);
+			let Some(valid_until) = LegacyLease::get_parachain_lease_in_blocks(p) else {
+				log::error!("Lease holding chain with no lease information?!");
+				return None
+			};
+			let valid_until: u32 = match valid_until.try_into() {
+				Ok(val) => val,
+				Err(_) => {
+					log::error!("Converting block number to u32 failed!");
+					return None
+				},
+			};
+			// We assume the coretime chain set this parameter to the recommended value in RFC-1:
+			const TIME_SLICE_PERIOD: u32 = 80;
+			let round_up = if valid_until % TIME_SLICE_PERIOD > 0 { 1 } else { 0 };
+			let time_slice = valid_until / TIME_SLICE_PERIOD + TIME_SLICE_PERIOD * round_up;
+			log::trace!(target: "coretime-migration", "Sending of lease holding para {:?}, valid_until: {:?}, time_slice: {:?}", p, valid_until, time_slice);
+			Some(mk_coretime_call::<T>(CoretimeCalls::SetLease(p.into(), time_slice)))
+		});
+
+	let core_count: u16 = configuration::ActiveConfig::<T>::get()
```

Comment:
Hmm. on another call in a minute, but please check that this core handling is actually correct. This core_count value should read 0 on Polkadot at the time of the migration.

Reply:
It's not 0. We update it on line 211-214:
https://github.com/polkadot-fellows/runtimes/pull/426/files#diff-6a311f9b6c4cd2a09f397e5e0c90ea9193ba103bf277bb88de6bd223148af52eR211-R214

The actual value is 52:
```
coretime-migration  TRACE: Set core count to 52. legacy paras count is  52
```

The reserved cores are 0 though because `legacy_paras_count == core_count`. I have got an assert for this in the test:
https://github.com/tdimitrov/polkadot-coretime-launch-test/blob/a8e7aba7d6a741b10bd5330ae3119ef15a018711/coretime-migration-check/main.js#L158-L160

Is this correct?

Reply:
> The reserved cores are 0 though

What you mean by reserved cores? The system chains? 

Generally your code is correct.

Reply:
Yes, the system chains.

Reply:
No, wait. These are reserved for on demand, if I am not mistaken.
We reserve cores for system chains, then for lease cores and what's left is for on demand. Right @eskimor ?